Electric service refactoring followup work

EnergyArchmage opened the pull request:
Address comments, completed remaining items from  https://github.com/NREL/EnergyPlus/issues/5499

EnergyArchmage committed:
generator list getObjectItem call

fix mistake in GetObjectItem call for generator list.  add trap for
input error for missing transformer,

EnergyArchmage committed:
And eng ref Changes

clean up end use subcats.  missing internal gain keys in IDD

EnergyArchmage committed:
1st pass at C++ cleanup

addressing comments from SM, remove private default constructors, moving
member inits into main constructors.  Add trailing "_" to private data

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 5499-Electric-Service-Follow-Up

EnergyArchmage committed:
remove "this->"

EnergyArchmage committed:
string arguments passed by reference

EnergyArchmage committed:
Clean up comments and whitespace

EnergyArchmage committed:
fix storage reinit at end warmup

EnergyArchmage committed:
address review comments in Eng Ref

comments on eng ref doc from Kandler Smith, battery expert from NREL's
Transportation Center

EnergyArchmage committed:
Debug internal gains, expand input checks

Fix internal gains omissions, add more complex integration file,  minor
corrections to testfiles.  More robust checks of devices on multiple
load centers.

DeadParrot commented on src/EnergyPlus/ElectricPowerServiceManager.hh:
```diff
@@ -162,6 +124,8 @@ public: // Methods
 	getACEnergyOut();
 
+	std::string
+	name();
```
You can return by const reference to avoid expensive string copy since `name()` is just returning an existing string. Methods that don't modify the object should be declared `const`.

```
std::string const &
name() const;
```

Same comment for all the `name()` methods below.

DeadParrot commented:
@EnergyArchmage Looks like you hit all the items we discussed. Looking at the code I notice a few more things (sorry!)...
-  Methods like `name()` that don't modify the object should be declared const so that they can be called on an instance that is const in the local context. This includes all the `name()` methods, `transformerName()`, the `get*()` methods, and so on.
-  "Getters" like `name()` and `get*()` should return "big" (non-POD) values like `std::string` by const reference to avoid an unnecessary creation/copy/heap hit. Just declare the return type like `std::string const &` .
-  You can remove the explicit destructors that don't do anything if you want: C++ will do the automatic "destruct every member" destructor for you. Once you have inheritance then you might need empty destructors, at least in the base class, so you can declare them virtual. And you need to write destructors if they have to do things like releasing resources that the C++ automatic destructor won't do.
- Consider having a method like `calcLoadCenterThermalLoad` return the computed value instead of passing it as a reference since it doesn't use the incoming value -- you get that information at the call site and it lets you write the call more compactly.
-  I see a block of public data commented as being (public) for the unit test benefit. If that is true then it is something that might be worth another look. In a more rigorous development setting that would not be considered a good reason for public data (and public data would probably be disallowed except in extreme cases). I guess the philosophy is you provide enough API to test the class from its public interface only without looking at internal private state. The problem with that is wasted time writing tons of little getter methods. Since we are far from proper OO anyway I leave it to you to decide this one.

Some minor, non-essential cosmetic stuff:
- Constructors before destructors feels more natural when reading code.
- Normally there is no space between a type name and the < for the template arg: `std::vector< int >`
- It is OK and probably nicer to put single arguments in the same line with the function name in declarations (unless you really want a trailing comment for that argument): `ElectPowerLoadCenter( int const objectNum );`
- There are a few mis-indentings and whitespace issues that would be nice to make a pass for. Some of these are:
  - Extra indentation of the constructor member initialization sections (probably from cut and paste)
  - I generally suggest not adding extra whitespace to align the = or things like that: they look nice at first but are rarely maintained as longer items are added and cause extra diff lines if someone does maintain them.
  - A consistent method declaration format with the return type on its own line is suggested: you've got some of each now.
  - I suggest single blank lines around the `public:` and `private:` lines and normally no more than a single blank line between functions or lines of code.
  - The objects and functions outside the class and just in the EnergyPlus namespace should be unindented to column 1 since we aren't indenting for namespaces:
- Extraneous parentheses should be removed for readability and because some compilers generate warnings about them at some level of pickiness: `if ( storOpCVGenRate < ( adjustedFeedInRequest ) ) {`

EnergyArchmage committed:
Addressing latest comments

many changes to c++ style.  Nearly all addressed except for data made
public for unit tests.

EnergyArchmage commented:
Okay, thanks @DeadParrot .  I think I have addressed these comments, except for the data made public in class ElectPowerLoadCenter for unit testing.  

Note that the test failures here are expected because this PR includes some changes/fixes, especially to reset electric storage to initial state of charge at end of warmup.

DeadParrot commented:
I think the C++ is in good shape. Some very minor notes:
- Using const on declarations when appropriate is a good idea, like:

``` cpp
Real64 const loadCenterThermalLoad = remainingThermalLoad;
```
- A comment at the top of each class describing its purpose would be nice

The remaining public data use may be worth reassessing if it is for expedience: the use of friends and some additional API might allow it to be eliminated. Ideally, internal state is protected against monkeying with outside of the class' interface to reduce bug exposure. Since EnergyPlus isn't clean OO yet it isn't clear how strict we should be about this for new code.

Note that I didn't review wrt functionality, unit test coverage, CI diffs, etc.

EnergyArchmage committed:
add some class level comments

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 5499-Electric-Service-Follow-Up

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into 5499-Electric-Service-Follow-Up

DeadParrot commented:
The C++ usage issues I noticed have been suitably addressed to move forward with this once the diffs and any other concerns are dealt with.

Myoldmopar committed:
Port documentation from MD to LaTeX

Myoldmopar committed:
Update engref CMakeLists for updated image names and new file

Myoldmopar commented:
@EnergyArchmage I finished the TeX changeover in this branch.  No conflicts.  I'll do some more review and testing as needed, and this should drop in.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -19458,6 +19458,7 @@ RoomAir:Node:AirflowNetwork:InternalGains,
       \key ElectricLoadCenter:Inverter:LookUpTable
       \key ElectricLoadCenter:Storage:Battery
       \key ElectricLoadCenter:Storage:Simple
+      \key ElectricLoadCenter:Storage:Converter
```
@EnergyArchmage Since we are past the IO lockdown, this has me concerned.  If this can't go in because of the IDD change here, is there anything you'd like to, or be able to, extract out as a separate, tiny, bug fix PR?  If not, this may have to wait until after release.  Let me get a feel from some interface folks here.

@kbenne @macumber Does this type of IDD change have an effect on you all?  This is a refactoring PR that preps the code for a separate bug fix PR (#5529) that would be nice to get in, but since it affects the idd here, I'm curious if this will have an effect on you downstream.

EnergyArchmage replied:
These IDD changes are no big deal to omit if you want.

Myoldmopar replied:
OK, I'll push a commit to this branch without those idd changes and push them up in a separate branch immediately after.  This will slide right in.

@kbenne @macumber I'm still curious about whether there is a list of acceptable changes/unacceptable changes to the IDD after the IO freeze.  If the freeze is 100% frozen except SHA, then so be it, just curious if there is a clean list.

macumber replied:
This addition is ok, it would be a new feature that openstudio has not yet wrapped.  The issue would be on translation from idd to osm, we need to put this in our e+ idd or else we could not open idf files with this field.  So it is ok but we do need to know about it.  Is this going in?

Myoldmopar replied:
Thanks for writing back.  No, we pulled this out and will merge it in right after release.  No worries this time.

Myoldmopar committed:
Revert IDD changes back out

Myoldmopar commented:
OK, reverted the IDD changes back out.  Created a new branch off of develop to re-add those in post-release (#5569).  I can merge this in now.
